File: symfony/src/Service/Tournaments/TournamentRegistrationService.php
<?php

namespace App\Service\Tournaments;

use App\Entity\Tournaments\TournamentRegistration;
use App\Repository\Tournaments\TournamentRegistrationRepository;

class TournamentRegistrationService
{
    public function __construct(
        private TournamentRegistrationRepository $repository,
    ) {}

    public function create(array $data): TournamentRegistration
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TournamentRegistration $entity, array $data): TournamentRegistration
    {
        throw new \LogicException('Not implemented');
    }

    public function withdraw(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TournamentRegistration not found: ' . $id);
        $entity->withdraw();
        $this->repository->save($entity, flush: true);
    }

    public function disqualify(int $id, $reason): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TournamentRegistration not found: ' . $id);
        $entity->disqualify($reason);
        $this->repository->save($entity, flush: true);
    }

    public function promoteFromWaitlist(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TournamentRegistration not found: ' . $id);
        $entity->promoteFromWaitlist();
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/tournaments/tournament_registration_service.rb
module Tournaments
  class TournamentRegistrationService
    def initialize(repository = TournamentRegistration)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def withdraw(id)
      instance = TournamentRegistration.find(id)
      instance.withdraw()
      instance.save!
    end

    def disqualify(id, reason)
      instance = TournamentRegistration.find(id)
      instance.disqualify(reason)
      instance.save!
    end

    def promote_from_waitlist(id)
      instance = TournamentRegistration.find(id)
      instance.promote_from_waitlist()
      instance.save!
    end
  end
end
